### PR TITLE
2.0

### PR DESCRIPTION
**Pull request by jpatel531**
Cleaner, more testable + more intuitive to use

### CODE REVIEW

**Comment by jpatel531** [comment]
@vivangkumar you wanna review b11657a ?

File: notifications.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	Apns         *ApnsNotification `json:"apns,omitempty"`
+	Gcm          *GcmNotification  `json:"gcm,omitempty"`
+}
+
+type ApnsNotification struct {
+	Id         string       `json:"apns-id,omitempty"`
+	Expiration int64        `json:"expiration,omitempty"`
+	Priority   int          `json:"priority,omitempty"`
+	CollapseID string       `json:"collapse_id,omitempty"`
+	Payload    *ApnsPayload `json:"aps"`
```
Comment (topliceanu) [inline notifications.go]: AFAIK, there should also be a `"data"` key in this structure!

File: pusher.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+		host:    p.Host,
+		cluster: p.Cluster,
+		appID:   p.appID,
+		secure:  p.Secure,
+	}
+}
+
+func (p *Pusher) httpClient() *http.Client {
+	if p.HttpClient == nil {
+		p.HttpClient = &http.Client{Timeout: time.Second * 5}
```
Comment (topliceanu) [inline pusher.go]: I think this timeout should be configurable by the library's clients, at least that's the case for the other languages.

File: pusher.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	"github.com/pusher/pusher-http-go/requests"
+	"github.com/pusher/pusher-http-go/signatures"
+	"github.com/pusher/pusher-http-go/validate"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+const (
+	NATIVE_NOTIFICATIONS_HOST = "yolo.ngrok.io"
```
Comment (topliceanu) [inline pusher.go]: @jpatel531 it took me some time to realize I was making calls to your localhost :) Shouldn't this be configurable in `Options`?

Comment (vivangkumar) [inline pusher.go]: Hahahaha, it strikes again

**Comment by sparrc** [comment]
is this a breaking change for users? do you have any idea when it may get merged?

curious because I have an open PR which uses this client library in https://github.com/influxdata/telegraf

thanks!

**Comment by jpatel531** [comment]
Hey there @sparrc - at the moment this is a breaking change. But I've been thinking there's no real pressing need for there to be, so I'm keen to refactor it some more to make it backwards compatible. Thanks for pointing this out. Currently I haven't had the time to work much on it (except for the odd fix where we're using this version of the library internally), but maybe in the next couple of weeks I can give it some attention and release it.

**Comment by sparrc** [comment]
Okay, thanks for the info!

For us I don't mind if it's a breaking change. We vendor all of our dependencies so we won't be affected until we update the dependency by choice (at which point we can deal with the refactor).

When it's happened in the past it's usually been fairly straight-forward, if it isn't then just a little blurb on the readme usually helps :)

**Comment by vivangkumar** [comment]
I'm going to take this over and try to merge it in. Looking to get some work done on it the remaining of the week. @jpatel531 are there things I should take note of?

**Comment by jpatel531** [comment]
Mmm not really. I think i was a bit too flair with some of the abstractions and sub packages. Just go wild and do what you think is right.

**Comment by vivangkumar** [comment]
Just an update: Ive been working on this whenever I've had time. I can hopefully push some a few changes I made soon. Most of the work is the same, just some minor improvements here and there.

**Comment by jpatel531** [comment]
DENIED!!!!

**Comment by kn100** [comment]
Is this still relevant?

**Comment by topliceanu** [comment]
It's up to you @kn100. This work is very old, if you can salvage anything, go ahead!

**Comment by jpatel531** [comment]
yeah i would just close this. now i look back on it it's quite unnecessarily verbose and the fact it's split into packages only used internally could be annoying for customers' dependency management tools

**Comment by kn100** [comment]
Closing, I'll look at this when I get some time to see if there's anything worth salvaging.

**Pull request closed without merging**